dtfs-central-api/src/services/utilisation-report-data-validator/utilisation-report-cell-validators/helpers/generate-errors-for-mismatched-facility-values/index.ts

```diff
@@ -0,0 +1,54 @@
+import { UtilisationReportCsvRowData, UtilisationReportDataValidationError } from '@ukef/dtfs2-common';
+import objectsAreEqual from '../../../../../helpers/objects-are-equal';
+
+/**
+ * generateErrorsForMismatchedFacilityValues
+ * Adds error message for all matching facility id rows
+ * Finds all rows with the same facility id
+ * Generates errors if errors do not already exist in the provided errors object t
+ * @param {UtilisationReportCsvRowData[]} csvData
+ * @param {UtilisationReportDataValidationError[]} errors - existing validation errors
+ * @param {UtilisationReportCsvRowData} csvRow - current CSV row to generate errors from
+ * @param {String} field - field to obtain values
+ * @param {String} errorMessage - error message to generate
+ * @param {String} exporterName - exporter name for errors
+ * @returns {UtilisationReportDataValidationError[]} generated errors
+ */
+const generateErrorsForMismatchedFacilityValues = (
+  csvData: UtilisationReportCsvRowData[],
+  errors: UtilisationReportDataValidationError[],
+  csvRow: UtilisationReportCsvRowData,
+  field: string,
+  errorMessage: string,
+  exporterName: string,
+) => {
+  // find all rows with matching facility id
+  const matchingRows = csvData.filter((row: UtilisationReportCsvRowData) => row['ukef facility id']?.value === csvRow['ukef facility id']?.value);
+  const newErrors: UtilisationReportDataValidationError[] = [];
+
+  /**
+   * iterate through rows with same facility id
+   * generate error message from row and provided error message
+   * check if row contains identical error
+   * if the error is unique, push the error to the errors array
+   */
+  matchingRows.forEach((row: UtilisationReportCsvRowData) => {
+    const errorObject = {
+      errorMessage,
+      column: row[field]?.column,
+      row: row[field]?.row,
+      value: row[field]?.value,
+      exporter: exporterName,
+    } as UtilisationReportDataValidationError;
+
+    const isNotInErrors = !errors.some((error) => objectsAreEqual(error, errorObject));
```

Comment:
How would we be generating identical error messages if we are only looping over the array once?

Reply:
We go through the sheet row by row, so if we have 3 rows, the error will be generated once per loop for each row, so we would end up with 9 error messages for the same rows

Reply:
Ah, if we're doing that why don't we just add the errors for the row we are iterating on at that point instead of adding all 3 3 times and then needing to remove some?
We're already doing some fairly complex O(n^2) logic so not double counting then needing to remove duplicates would be great

Reply:
we aren't removing any, we're just not adding them if they exist.  I do not know any other way to check row by row and add errors for each row which has the same facility id

Reply:
Ah right I think I get it, misunderstood before! In that case could we just use Lodash's isEqual helper here instead of writing our own objectsAreEqual helper please? It's covers a lot more cases than what our one does and means we don't need to maintain it 
https://www.geeksforgeeks.org/lodash-_-isequal-method/